### PR TITLE
Elide Block null arrays if all values non-null

### PR DESCRIPTION
**Pull request by dain**
For fixed witdth data this saves ~10% of memory

### CODE REVIEW

**Comment by dain** [comment]
This is a demo of an idea to reduce memory usage when blocks have no nulls (the normal case).  If this is acceptable, I'll apply the same change to all blocks.

cc @findepi, @sopel39

File: presto-thrift-connector-api/src/main/java/com/facebook/presto/connector/thrift/api/datatypes/PrestoThriftBigint.java [hunk 1]
```diff
@@ -77,7 +78,7 @@ public Block toBlock(Type desiredType)
         int numberOfRecords = numberOfRecords();
         return new LongArrayBlock(
                 numberOfRecords,
-                nulls == null ? new boolean[numberOfRecords] : nulls,
+                Optional.ofNullable(nulls),
```
Comment (electrum) [inline presto-thrift-connector-api/src/main/java/com/facebook/presto/connector/thrift/api/datatypes/PrestoThriftBigint.java]: The Thrift contract and original code seem a bit broken for the case of all nulls. If both the `nulls` and `longs` arrays are elided / null, I would expect that you get all nulls. However, the current code gives you `0` for every row. If we are fine clarifying the contract to cover this case, then we can use an RLE block.

**Comment by electrum** [comment]
Looks good to me. I like this idea.

**Comment by dain** [comment]
This is ready for review now.

File: presto-main/src/test/java/com/facebook/presto/block/AbstractTestBlock.java [hunk 2]
```diff
@@ -108,7 +108,8 @@ private void assertRetainedSize(Block block)
                 field.setAccessible(true);
 
                 if (type == Slice.class) {
-                    retainedSize += ((Slice) field.get(block)).getRetainedSize();
+                    Slice slice = (Slice) field.get(block);
+                    retainedSize += slice == null ? 0 : slice.getRetainedSize();
```
Comment (findepi) [inline presto-main/src/test/java/com/facebook/presto/block/AbstractTestBlock.java]: nit: 
```
if (slice !=null) {
 retainedSize += slice.getRetainedSize();
}
```

File: presto-spi/src/main/java/com/facebook/presto/spi/block/FixedWidthBlock.java [hunk 3]
```diff
@@ -76,20 +91,22 @@ public int getPositionCount()
     @Override
     public long getSizeInBytes()
     {
-        return getRawSlice().length() + (long) valueIsNull.length();
+        return (fixedSize + Byte.BYTES) * (long) positionCount;
```
Comment (findepi) [inline presto-spi/src/main/java/com/facebook/presto/spi/block/FixedWidthBlock.java]: Extract to a separate commit, since adhering to `getSizeInBytes` contract has nothing to do with eliding the `valueIsNull` array

File: presto-spi/src/main/java/com/facebook/presto/spi/block/LongArrayBlockBuilder.java [hunk 4]
```diff
@@ -87,6 +88,7 @@ public BlockBuilder appendNull()
 
         valueIsNull[positionCount] = true;
 
+        hasNullValue = true;
```
Comment (sopel39) [inline presto-spi/src/main/java/com/facebook/presto/spi/block/LongArrayBlockBuilder.java]: Shouldn't `valueIsNull` be initialized on first null to reduce builder memory footprint?

**Comment by sopel39** [comment]
It seems that some tests legitimately fail

File: presto-spi/src/main/java/com/facebook/presto/spi/block/LongArrayBlockBuilder.java (hunk 4 above)
Comment (dain) [reply presto-spi/src/main/java/com/facebook/presto/spi/block/LongArrayBlockBuilder.java]: Yes that could be done also.  Maybe some good followup work.

File: presto-spi/src/main/java/com/facebook/presto/spi/block/Block.java [hunk 5]
```diff
@@ -239,6 +239,17 @@ default Block getPositions(int[] positions, int offset, int length)
      */
     Block copyRegion(int position, int length);
 
+    /**
+     * Is it possible the block may have nulls?  If false, the block can not have
+     * any nulls, but if true, the block may or may not have nulls.
+     *
+     * @throws IllegalArgumentException if this position is not valid
+     */
+    default boolean mayHaveNull()
```
[the remark this answers is not in the document]
Comment (dain) [reply presto-spi/src/main/java/com/facebook/presto/spi/block/Block.java]: Change docs to singular
